In [16]:
# Importing the required libraries 
import os
import numpy as np
import pandas as pd
import datetime as dt
import seaborn as sns
import matplotlib.pyplot as plt
import glob
from pyculiarity import detect_ts

# Load Meter and Meta Data

In [2]:
all_meters = pd.read_csv("../data/meters/cleaned/allmeters_daily.csv")
meta_data = pd.read_csv("../data/metadata/metadata.csv")

We are only interested looking at Energy Consumption by 
- Educational Institutions
- Office

### Process metadata buiding

In [3]:
#meta_education = meta_data[meta_data['primaryspaceusage'] == 'Education']
# meta_education
meta_education_office = meta_data[meta_data['primaryspaceusage'].isin(['Education', 'Office'])]
meta_education_office

building_id   site_id  building_id_kaggle  \
3       Panther_education_Rosalie   Panther                 0.0   
4         Panther_education_Misty   Panther                 1.0   
5            Panther_office_Daina   Panther                 NaN   
6        Panther_education_Mattie   Panther                 2.0   
7          Panther_office_Woodrow   Panther                 NaN   
...                           ...       ...                 ...   
1619    Cockatoo_education_Collin  Cockatoo              1439.0   
1621    Cockatoo_education_Gussie  Cockatoo              1441.0   
1623     Cockatoo_education_Arlen  Cockatoo              1443.0   
1625  Cockatoo_education_Victoria  Cockatoo              1445.0   
1628     Cockatoo_office_Margaret  Cockatoo              1448.0   

      site_id_kaggle primaryspaceusage sub_primaryspaceusage     sqm     sqft  \
3                0.0         Education              Research   690.5   7432.0   
4                0.0         Education              Research   252.7   2720.0   
5                0.0            Office                Office   133.8   1440.0   
6                0.0         Education             Classroom   499.4   5376.0   
7                0.0            Office                Office   537.8   5789.0   
...              ...               ...                   ...     ...      ...   
1619            15.0         Education    College Laboratory  5248.7  56497.0   
1621            15.0         Education    College Laboratory  2800.4  30143.0   
1623            15.0         Education     College Classroom  3745.0  40311.0   
1625            15.0         Education     College Classroom   399.3   4298.0   
1628            15.0            Office                Office  8572.3  92271.0   

            lat        lng  ... yearbuilt date_opened numberoffloors  \
3     28.517689 -81.379039  ...    2008.0         NaN            NaN   
4     28.517689 -81.379039  ...    2004.0         NaN            NaN   
5     28.517689 -81.379039  ...    1983.0         NaN            NaN   
6     28.517689 -81.379039  ...    1991.0         NaN            NaN   
7     28.517689 -81.379039  ...    1996.0         NaN            NaN   
...         ...        ...  ...       ...         ...            ...   
1619  42.459837 -76.485292  ...    1957.0         NaN            NaN   
1621  42.459837 -76.485292  ...    1951.0         NaN            NaN   
1623  42.459837 -76.485292  ...    1913.0         NaN            NaN   
1625  42.459837 -76.485292  ...       NaN         NaN            NaN   
1628  42.459837 -76.485292  ...    2001.0         NaN            NaN   

     occupants energystarscore   eui site_eui source_eui leed_level rating  
3          NaN             NaN   276      NaN        NaN        NaN    NaN  
4          NaN             NaN   375      NaN        NaN        NaN    NaN  
5          NaN             NaN  1983      NaN        NaN        NaN    NaN  
6          NaN             NaN    49      NaN        NaN        NaN    NaN  
7          NaN             NaN   136      NaN        NaN        NaN    NaN  
...        ...             ...   ...      ...        ...        ...    ...  
1619       NaN             NaN   NaN      NaN        NaN        NaN    NaN  
1621       NaN             NaN   NaN      NaN        NaN        NaN    NaN  
1623       NaN             NaN   NaN      NaN        NaN        NaN    NaN  
1625       NaN             NaN   NaN      NaN        NaN        NaN    NaN  
1628       NaN             NaN   NaN      NaN        NaN        NaN    NaN  

[924 rows x 32 columns]

We have 924 buildings that were used for Educational and Office purposes

In [4]:
meta_education_office.info()

<class 'pandas.core.frame.DataFrame'>
Index: 924 entries, 3 to 1628
Data columns (total 32 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   building_id            924 non-null    object 
 1   site_id                924 non-null    object 
 2   building_id_kaggle     828 non-null    float64
 3   site_id_kaggle         915 non-null    float64
 4   primaryspaceusage      924 non-null    object 
 5   sub_primaryspaceusage  924 non-null    object 
 6   sqm                    924 non-null    float64
 7   sqft                   924 non-null    float64
 8   lat                    827 non-null    float64
 9   lng                    827 non-null    float64
 10  timezone               924 non-null    object 
 11  electricity            900 non-null    object 
 12  hotwater               117 non-null    object 
 13  chilledwater           371 non-null    object 
 14  steam                  235 non-null    object 
 15  water     

Base on the EDA, and the number of missing values, the following fields in metadata will be kept:
1. building_id
2. site_id
3. primaryspaceusage
4. sqm
5. sqft
6. timezone
7.  



In [5]:
# Since not all buildings use all energy sources, marking the sources used for each building 
meter_columns = ['electricity', 'hotwater', 'chilledwater', 'steam', 'water', 'irrigation', 'solar', 'gas']
for col in meter_columns:
    meta_education_office[col] = np.where(meta_education_office[col] == 'Yes', 1, 0)   # where the meter is present, mark 1, else 0

/tmp/ipykernel_32197/118997712.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  meta_education_office[col] = np.where(meta_education_office[col] == 'Yes', 1, 0)   # where the meter is present, mark 1, else 0


In [6]:
# Melt the DataFrame to pivot it
meta_melted = meta_education_office.melt(id_vars=[col for col in meta_education_office.columns if col not in meter_columns], 
                    value_vars=meter_columns, 
                    var_name='meter', 
                    value_name='value')

# Filtering for the corrrect 
meta_melted = meta_melted[meta_melted['value']==1]

In [7]:
# Merging both the dataframes together
merged_df = pd.merge(all_meters, meta_melted, on=['building_id','meter' ], how='inner')

In [8]:
merged_df.columns

Index(['building_id', 'meter', 'date', 'meter_reading', 'site_id',
       'building_id_kaggle', 'site_id_kaggle', 'primaryspaceusage',
       'sub_primaryspaceusage', 'sqm', 'sqft', 'lat', 'lng', 'timezone',
       'industry', 'subindustry', 'heatingtype', 'yearbuilt', 'date_opened',
       'numberoffloors', 'occupants', 'energystarscore', 'eui', 'site_eui',
       'source_eui', 'leed_level', 'rating', 'value'],
      dtype='object')

In [9]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1343578 entries, 0 to 1343577
Data columns (total 28 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   building_id            1343578 non-null  object 
 1   meter                  1343578 non-null  object 
 2   date                   1343578 non-null  object 
 3   meter_reading          1343578 non-null  float64
 4   site_id                1343578 non-null  object 
 5   building_id_kaggle     1238314 non-null  float64
 6   site_id_kaggle         1334075 non-null  float64
 7   primaryspaceusage      1343578 non-null  object 
 8   sub_primaryspaceusage  1343578 non-null  object 
 9   sqm                    1343578 non-null  float64
 10  sqft                   1343578 non-null  float64
 11  lat                    1144746 non-null  float64
 12  lng                    1144746 non-null  float64
 13  timezone               1343578 non-null  object 
 14  industry          

In [10]:
merged_df.sample(10)

building_id        meter        date  meter_reading  \
272171    Cockatoo_education_Doreen  electricity  2016-08-27      2188.6750   
1212964          Rat_education_Mose  electricity  2016-08-23      3526.7900   
921603         Lamb_education_Patsy  electricity  2017-06-27         0.0072   
874689     Lamb_education_Gabrielle  electricity  2017-02-17       231.1010   
122020         Bull_education_Bryon        steam  2017-11-05        11.1630   
1078601        Panther_office_Daina        water  2017-01-11       113.5623   
850063          Lamb_education_Chet  electricity  2017-10-03         0.0024   
501184         Eagle_office_Phyllis     hotwater  2017-03-25    348752.5851   
300306   Cockatoo_education_Latrice  electricity  2017-08-19         0.6750   
1248926             Rat_office_Jill  electricity  2017-01-13     21423.2600   

          site_id  building_id_kaggle  site_id_kaggle primaryspaceusage  \
272171   Cockatoo              1356.0            15.0         Education   
1212964       Rat               412.0             3.0         Education   
921603       Lamb                 NaN             5.0         Education   
874689       Lamb               706.0             5.0         Education   
122020       Bull               891.0             9.0         Education   
1078601   Panther                 NaN             0.0            Office   
850063       Lamb                 NaN             5.0         Education   
501184      Eagle              1297.0            14.0            Office   
300306   Cockatoo              1345.0            15.0         Education   
1248926       Rat               560.0             3.0            Office   

               sub_primaryspaceusage      sqm  ...  date_opened  \
272171             College Classroom   8477.7  ...          NaN   
1212964                  K-12 School  10219.3  ...          NaN   
921603   Primary/Secondary Classroom   2991.0  ...          NaN   
874689   Primary/Secondary Classroom   2572.0  ...          NaN   
122020             College Classroom   4723.7  ...          NaN   
1078601                       Office    133.8  ...          NaN   
850063   Primary/Secondary Classroom   1078.0  ...          NaN   
501184                        Office   1994.5  ...          NaN   
300306            College Laboratory   1940.7  ...          NaN   
1248926                       Office  31587.0  ...          NaN   

         numberoffloors  occupants energystarscore   eui site_eui source_eui  \
272171              NaN        NaN             NaN   NaN      NaN        NaN   
1212964             NaN        NaN             NaN   NaN      NaN        NaN   
921603              2.0      583.0             NaN   NaN      NaN        NaN   
874689              2.0      414.0             NaN   NaN      NaN        NaN   
122020              NaN        NaN             NaN   NaN      NaN        NaN   
1078601             NaN        NaN             NaN  1983      NaN        NaN   
850063              2.0      235.0             NaN   NaN      NaN        NaN   
501184              NaN        NaN             NaN   NaN      NaN        NaN   
300306              NaN        NaN             NaN   NaN      NaN        NaN   
1248926             NaN        NaN             NaN   NaN      NaN        NaN   

         leed_level rating  value  
272171          NaN    NaN      1  
1212964         NaN    NaN      1  
921603          NaN      D      1  
874689          NaN      D      1  
122020          NaN    NaN      1  
1078601         NaN    NaN      1  
850063          NaN      C      1  
501184          NaN    NaN      1  
300306          NaN    NaN      1  
1248926         NaN    NaN      1  

[10 rows x 28 columns]

In [11]:
# Inspecting the dataframe for a single building id 
merged_df[merged_df['building_id_kaggle'] == 636]

building_id        meter        date  meter_reading site_id  \
0    Bear_education_Alfredo  electricity  2016-01-01         2.9050    Bear   
1    Bear_education_Alfredo  electricity  2016-01-02         2.7700    Bear   
2    Bear_education_Alfredo  electricity  2016-01-03         2.6725    Bear   
3    Bear_education_Alfredo  electricity  2016-01-04         4.5650    Bear   
4    Bear_education_Alfredo  electricity  2016-01-05         4.7825    Bear   
..                      ...          ...         ...            ...     ...   
726  Bear_education_Alfredo  electricity  2017-12-27         3.9825    Bear   
727  Bear_education_Alfredo  electricity  2017-12-28         3.9975    Bear   
728  Bear_education_Alfredo  electricity  2017-12-29         3.3650    Bear   
729  Bear_education_Alfredo  electricity  2017-12-30         3.1825    Bear   
730  Bear_education_Alfredo  electricity  2017-12-31         3.1800    Bear   

     building_id_kaggle  site_id_kaggle primaryspaceusage  \
0                 636.0             4.0         Education   
1                 636.0             4.0         Education   
2                 636.0             4.0         Education   
3                 636.0             4.0         Education   
4                 636.0             4.0         Education   
..                  ...             ...               ...   
726               636.0             4.0         Education   
727               636.0             4.0         Education   
728               636.0             4.0         Education   
729               636.0             4.0         Education   
730               636.0             4.0         Education   

    sub_primaryspaceusage    sqm  ...  date_opened  numberoffloors  occupants  \
0               Education  609.8  ...          NaN             3.0        NaN   
1               Education  609.8  ...          NaN             3.0        NaN   
2               Education  609.8  ...          NaN             3.0        NaN   
3               Education  609.8  ...          NaN             3.0        NaN   
4               Education  609.8  ...          NaN             3.0        NaN   
..                    ...    ...  ...          ...             ...        ...   
726             Education  609.8  ...          NaN             3.0        NaN   
727             Education  609.8  ...          NaN             3.0        NaN   
728             Education  609.8  ...          NaN             3.0        NaN   
729             Education  609.8  ...          NaN             3.0        NaN   
730             Education  609.8  ...          NaN             3.0        NaN   

    energystarscore  eui site_eui source_eui  leed_level rating  value  
0               NaN  NaN      NaN        NaN         NaN    NaN      1  
1               NaN  NaN      NaN        NaN         NaN    NaN      1  
2               NaN  NaN      NaN        NaN         NaN    NaN      1  
3               NaN  NaN      NaN        NaN         NaN    NaN      1  
4               NaN  NaN      NaN        NaN         NaN    NaN      1  
..              ...  ...      ...        ...         ...    ...    ...  
726             NaN  NaN      NaN        NaN         NaN    NaN      1  
727             NaN  NaN      NaN        NaN         NaN    NaN      1  
728             NaN  NaN      NaN        NaN         NaN    NaN      1  
729             NaN  NaN      NaN        NaN         NaN    NaN      1  
730             NaN  NaN      NaN        NaN         NaN    NaN      1  

[731 rows x 28 columns]

# Missing Data

In [12]:
def determine_column_counts(df):
    columns = df.columns
    data_counts = pd.DataFrame(columns = ['column_name','column_type','num_rows','distinct_data', 'distinct_percent', 'missing_data', 'missing_percentage'])
    
    #For each column, calculate number and percentage of missing rows and distinct counts
    for col in columns:
        col_vals = df[col]
        total_data = len(col_vals)
        col_missing = sum(col_vals.isnull())
        missing_percent = round((col_missing / len(df[col])) * 100 , 3)
        col_distinct = (col_vals.nunique())
        distinct_percent = round((col_distinct / len(df[col])) * 100 , 3)

        data_counts.loc[len(data_counts)] = [col, col_vals.dtype, total_data, col_distinct, distinct_percent, col_missing, missing_percent]
        
    return(data_counts)

missing_data = determine_column_counts(merged_df)
missing_data

column_name column_type  num_rows  distinct_data  \
0             building_id      object   1343578            924   
1                   meter      object   1343578              8   
2                    date      object   1343578            731   
3           meter_reading     float64   1343578         874020   
4                 site_id      object   1343578             17   
5      building_id_kaggle     float64   1343578            828   
6          site_id_kaggle     float64   1343578             16   
7       primaryspaceusage      object   1343578              2   
8   sub_primaryspaceusage      object   1343578             15   
9                     sqm     float64   1343578            906   
10                   sqft     float64   1343578            906   
11                    lat     float64   1343578             14   
12                    lng     float64   1343578             14   
13               timezone      object   1343578              6   
14               industry      object   1343578              2   
15            subindustry      object   1343578              4   
16            heatingtype      object   1343578              8   
17              yearbuilt     float64   1343578            109   
18            date_opened      object   1343578              0   
19         numberoffloors     float64   1343578             17   
20              occupants     float64   1343578             97   
21        energystarscore      object   1343578             26   
22                    eui      object   1343578            157   
23               site_eui      object   1343578            100   
24             source_eui      object   1343578            101   
25             leed_level      object   1343578              2   
26                 rating      object   1343578              8   
27                  value       int64   1343578              1   

    distinct_percent  missing_data  missing_percentage  
0              0.069             0               0.000  
1              0.001             0               0.000  
2              0.054             0               0.000  
3             65.052             0               0.000  
4              0.001             0               0.000  
5              0.062        105264               7.835  
6              0.001          9503               0.707  
7              0.000             0               0.000  
8              0.001             0               0.000  
9              0.067             0               0.000  
10             0.067             0               0.000  
11             0.001        198832              14.799  
12             0.001        198832              14.799  
13             0.000             0               0.000  
14             0.000        771205              57.399  
15             0.000        771205              57.399  
16             0.001       1160097              86.344  
17             0.008        716380              53.319  
18             0.000       1343578             100.000  
19             0.001       1008049              75.027  
20             0.007       1135243              84.494  
21             0.002       1178372              87.704  
22             0.012       1070915              79.706  
23             0.007       1178372              87.704  
24             0.008       1178372              87.704  
25             0.000       1319455              98.205  
26             0.001       1175448              87.486  
27             0.000             0               0.000

- On the basis of missing values and domain knowledge
- we shall drop the following columns since they are not directly consequential to our analysis

In [13]:
columns_to_drop = ['lat', 'lng','industry', 'subindustry',
                   'heatingtype', 'yearbuilt', 'date_opened', 'numberoffloors',
                   'occupants', 'energystarscore', 'eui', 'site_eui', 
                   'source_eui','leed_level', 'rating', 'value', 'site_id_kaggle', 'building_id_kaggle']

cleaned_df = merged_df.drop(columns = columns_to_drop)

In [14]:
cleaned_df.head()

building_id        meter        date  meter_reading site_id  \
0  Bear_education_Alfredo  electricity  2016-01-01         2.9050    Bear   
1  Bear_education_Alfredo  electricity  2016-01-02         2.7700    Bear   
2  Bear_education_Alfredo  electricity  2016-01-03         2.6725    Bear   
3  Bear_education_Alfredo  electricity  2016-01-04         4.5650    Bear   
4  Bear_education_Alfredo  electricity  2016-01-05         4.7825    Bear   

  primaryspaceusage sub_primaryspaceusage    sqm    sqft    timezone  
0         Education             Education  609.8  6564.0  US/Pacific  
1         Education             Education  609.8  6564.0  US/Pacific  
2         Education             Education  609.8  6564.0  US/Pacific  
3         Education             Education  609.8  6564.0  US/Pacific  
4         Education             Education  609.8  6564.0  US/Pacific

In [15]:
missing_data_after_cleaning = determine_column_counts(cleaned_df)   
missing_data_after_cleaning

column_name column_type  num_rows  distinct_data  \
0            building_id      object   1343578            924   
1                  meter      object   1343578              8   
2                   date      object   1343578            731   
3          meter_reading     float64   1343578         874020   
4                site_id      object   1343578             17   
5      primaryspaceusage      object   1343578              2   
6  sub_primaryspaceusage      object   1343578             15   
7                    sqm     float64   1343578            906   
8                   sqft     float64   1343578            906   
9               timezone      object   1343578              6   

   distinct_percent  missing_data  missing_percentage  
0             0.069             0                 0.0  
1             0.001             0                 0.0  
2             0.054             0                 0.0  
3            65.052             0                 0.0  
4             0.001             0                 0.0  
5             0.000             0                 0.0  
6             0.001             0                 0.0  
7             0.067             0                 0.0  
8             0.067             0                 0.0  
9             0.000             0                 0.0

In [17]:
os.makedirs("../data/cleaned", exist_ok=True)
cleaned_df.to_csv("../data/cleaned/cleaned_meters_meta.csv")

In [18]:
import gc
gc.collect()

0